http-client/Data/KeyedPool.hs

```diff
@@ -0,0 +1,316 @@
+    m <- atomically $ swapTVar var PoolClosed
+    F.mapM_ destroy m
+
+-- | Run a reaper thread, which will destroy old resources. It will
+-- stop running once our pool switches to PoolClosed, which is handled
+-- via the mkWeakIORef in the creation of the pool.
+reap :: forall key resource.
+        Ord key
+     => (resource -> IO ())
+     -> TVar (PoolMap key resource)
+     -> IO ()
+reap destroy var =
+    loop
+  where
+    loop = do
+        threadDelay (5 * 1000 * 1000)
+        join $ atomically $ do
+            m <- readTVar var
+            case m of
+                PoolClosed -> return (return ())
+                PoolOpen idleCount m
+                    | Map.null m -> retry
+                    | otherwise -> do
+                        (m', toDestroy) <- findStale idleCount m
+                        writeTVar var m'
+                        return $ do
+                            mask_ (mapM_ safeDestroy toDestroy)
+                            loop
+
+    findStale :: Int
+              -> Map key (PoolList resource)
+              -> STM (PoolMap key resource, [resource])
+    findStale idleCount m = do
+        -- We want to make sure to get the time _after_ any delays
+        -- occur due to the retry call above. Since getCurrentTime has
+        -- no side effects outside of the STM block, this is a safe
+        -- usage.
+        now <- unsafeIOToSTM getCurrentTime
+        let isNotStale time = 30 `addUTCTime` time >= now
+        let findStale' toKeep toDestroy [] =
+                (Map.fromList (toKeep []), toDestroy [])
+            findStale' toKeep toDestroy ((key, plist):rest) =
+                findStale' toKeep' toDestroy' rest
+              where
+                -- Note: By definition, the timestamps must be in
+                -- descending order, so we don't need to traverse the
+                -- whole list.
+                (notStale, stale) = span (isNotStale . fst) $ plistToList plist
+                toDestroy' = toDestroy . (map snd stale++)
+                toKeep' =
+                    case plistFromList notStale of
+                        Nothing -> toKeep
+                        Just x -> toKeep . ((key, x):)
+        let (toKeep, toDestroy) = findStale' id id (Map.toList m)
+        let idleCount' = idleCount - length toDestroy
+        return (PoolOpen idleCount' toKeep, toDestroy)
+
+    safeDestroy x = destroy x `catch` \(_ :: SomeException) -> return ()
+
+-- | Check out a value from the 'KeyedPool' with the given key.
+--
+-- This function will internally call 'mask_' to ensure async safety,
+-- and will return a value which uses weak references to ensure that
+-- the value is cleaned up. However, if you want to ensure timely
+-- resource cleanup, you should bracket this operation together with
+-- 'managedRelease'.
+takeKeyedPool :: Ord key => KeyedPool key resource -> key -> IO (Managed resource)
+takeKeyedPool kp key = mask_ $ join $ atomically $ do
```

Comment:
Without the `mask_`, an exception could be thrown between the completion of the `atomically` call and the completion of the `join` call, thereby preventing the `mkWeakIORef` call from ever being made.